index.js

```diff
@@ -25,15 +25,57 @@ function fastifyCors (fastify, opts, next) {
     allowedHeaders: null,
     maxAge: null,
     preflight: true,
-    hideOptionsRoute: true
+    hideOptionsRoute: true,
+    strictPreflight: false
   }, opts)
 
   const isOriginFalsy = !origin
   const isOriginString = typeof origin === 'string'
   const isOriginFunction = typeof origin === 'function'
 
   if (preflight === true) {
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => reply.send())
+    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+      if (origin === false) {
+        reply.code(404).type('text/plain').send('Not Found')
+        return
+      }
```

Comment:
I think we should not add the route if `!origin && preflight`.

Reply:
This is more complicated than I first realized.

In the original implementation, `origin` could be the result of a callback. As I see it, this means a few things:
1) My implementation here is not functionally equivalent to the previous implementation (I'm not calling the callback)
2) Makes not adding the route not feasible because if origin is a callback, it should be called on every request.

Reply:
Then call `origin`.

Reply:
I could call the origin callback in the hook and decorate the request with the result :thinking: Then check the value here and return 404 if it is false.

These options just keep getting more complicated :facepalm:

Reply:
Is it weird if it gets called twice during the same request? Once in the hook when adding the CORS headers and once in the route when adding the preflight stuff

Reply:
I could decorate the request and assign the result of the origin function to the request in the hook. That way it will only be called once per request. This is complicated but makes sense if the origin function is calling some external service or other weird behavior since it can return a promise.

Reply:
@mcollina I went with decorating the request with a boolean determining if the origin is allowed. This avoids calling a possibly asynchronous function (or promise) twice on every request.

Reply:
good job! I like this.